cmds/ocm/commands/controllercmds/install/cmd.go

```diff
@@ -0,0 +1,256 @@
+// SPDX-FileCopyrightText: 2022 SAP SE or an SAP affiliate company and Open Component Model contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package install
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/open-component-model/ocm/cmds/ocm/commands/controllercmds/names"
+	"github.com/open-component-model/ocm/cmds/ocm/commands/verbs"
+	"github.com/open-component-model/ocm/cmds/ocm/pkg/utils"
+	common2 "github.com/open-component-model/ocm/pkg/common"
+	"github.com/open-component-model/ocm/pkg/contexts/clictx"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+var (
+	Names = names.Controller
+	Verb  = verbs.Install
+)
+
+type Command struct {
+	utils.BaseCommand
+	Version       string
+	BaseURL       string
+	ReleaseAPIURL string
+}
+
+var _ utils.OCMCommand = (*Command)(nil)
+
+// NewCommand creates a new controller cdommand.
+func NewCommand(ctx clictx.Context, names ...string) *cobra.Command {
+	return utils.SetupCommand(&Command{BaseCommand: utils.NewBaseCommand(ctx)}, utils.Names(Names, names...)...)
+}
+
+func (o *Command) ForName(name string) *cobra.Command {
+	return &cobra.Command{
+		Use:   "install controller {--version v0.0.1}",
+		Short: "Install either a specific or latest version of the ocm-controller.",
+	}
+}
+
+func (o *Command) AddFlags(set *pflag.FlagSet) {
+	set.StringVarP(&o.Version, "version", "v", "latest", "the version of the controller to install")
+	set.StringVarP(&o.BaseURL, "base-url", "u", "https://github.com/open-component-model/ocm-controller/releases", "the base url to the ocm-controller's release page")
+	set.StringVarP(&o.ReleaseAPIURL, "release-api-url", "a", "https://api.github.com/repos/open-component-model/ocm-controller/releases", "the base url to the ocm-controller's API release page")
+}
+
+func (o *Command) Complete(args []string) error {
+	return nil
+}
+
+func (o *Command) Run() error {
```

Comment:
>Installing via shell execute has some drawbacks

What are some drawbacks?

>introduces a kind of hidden dependency.

You are installing a controller into a cluster. :) You probably already have a way of interacting with said cluster.

>Couldn't we use API calls here to perform the installation instead?

We could. It would complicate the code by a factor of 10. :)

Reply:
To be clear, I'm not opposed to using the API. And I guess it would add the option of adding potential values into the mix. For a first time implementation though, I'm not there is a distinctive benefit to doing so.

Reply:
Drawbacks are:

- You rely on the fact that user has installed kubectl. This is not always the case
- You rely on the fact that the kubectl executable is in the path
- You rely on the the fact that kubectl is installed in a compatible version. You don't have it under control
- If I install a CLI I assume that it is standalone tool and not relies on other components

I do not object against using shell execution but it is not nice. I would assume that there is a library that gives a similar abstraction as kubectl apply. Not sure if this is really factor 10.

Reply:
>You rely on the fact that user has installed kubectl. This is not always the 

Again, you are installing a thing into a cluster. You are supposed to have all tooling for said cluster.

> You rely on the fact that the kubectl executable is in the path

Same as above.

> You rely on the the fact that kubectl is installed in a compatible version. You don't have it under control

Doesn't matter what version it is as long as it can access the cluster.

> If I install a CLI I assume that it is standalone tool and not relies on other components

That is rarely the case. :) Also, we are depending on the fact that there is a cluster available and the user can interact with it. Otherwise they wouldn't be able to have a cluster in the first place.

>I do not object against using shell execution but it is not nice. I would assume that there is a library that gives a similar abstraction as kubectl apply. Not sure if this is really factor 10.

There is the k8s library for sure. But that would mean implementing order of apply, patching, and various other activities that require server side apply. For example the CRDs need to be installed first, then RBAC, etc.

I feel like that amount of work is not warranted for a such a small feature. :)

Reply:
We'll talk about this during today's sync. Let's find a nice middle ground. I absolutely share the concern of the outside CLI dependency.

Maybe we can iterate on this feature as time goes.

Reply:
Done. :)